### PR TITLE
Offer import when users click on supported import files

### LINKED ISSUE #1025: Offer import when users click on supported import files
We want to register Catima as an opener for common import file names so that users can easily start an import.

For Catima: catima.zip
For Fidme: fidme-export-request-*.zip
For Loyalty Card Keychain: LoyaltyCardKeychain.csv
For Stocard: *-sync.zip
For Voucher Vault: vouchervault.json

When the user opens such a file, Android should offer "Import data into Catima" as option. Catima should select the correct importer based on the filename.

### PR DESCRIPTION
**Pull request by TomerPacific**
Fixes #1025 

Logic was added to the **ImportExportActivity** to allow Catima to be associated with the following file types:

1. catima.zip 
2. fidme-export-request-*.zip
3.  LoyaltyCardKeychain.csv 
4.  *-sync.zip
5.  vouchervault.json

Due to limitations that were discussed in the issue, we are only able to associate the file extensions and not the file names themselves.

The **AndroidManifest.xml** has been augmented to hold intent filters for these file extensions:

- .zip
- .json
- .csv

196034131-6189e117-b0e9-41b7-8d7a-78c6d8287f31

### CODE REVIEW

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (TomerPacific) [inline app/src/main/java/protect/card_locker/ImportExportActivity.java:450]: @TheLastProject - I was looking for a constants file in the project and did not find one and wanted to know if you would prefer that I move the string literals here into constants?

Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java:450]: Hmm, I'm actually kinda worried with how this works. Opening a file will now instantly import with no user interaction, making it very easy to accidentally make a mess of everything. Maybe it's better to just open the list of supported importers and let the user choose (they'll likely know what kind of backup they're trying to import anyway). I think my idea of "Catima should select the correct importer based on the filename" was actually a bad idea seeing this.

I'm also a bit worried about the flow here as it's setting a variable that `openFileForImport` uses implicitly instead of something passed to the function explicitly. I see that that was part of the original design already though, so fixing that is probably out of scope (although probably should be done at some point, just like how the ImportExportActivity should be changed to first select a file and then the importer to prevent confusion for the use file manager and use other app being the same).

File: app/src/main/AndroidManifest.xml
Comment (TheLastProject) [inline app/src/main/AndroidManifest.xml:133]: Could you give the filters labels? Right now when I click a file Catima just says "Import/Export" when the only thing you can do in this flow is import. Maybe "Import cards" would be the best option

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (TheLastProject) [inline app/src/main/java/protect/card_locker/ImportExportActivity.java:77]: It probably shouldn't ask for storage permission when a file is shared to it (hmm, wonder if storage permission is even necessary nowadays and if it is which Android versions is is needed for, I need to look into #979 soon...)

**Review by TheLastProject — CHANGES_REQUESTED** [review]
Cool, works pretty well :)

File: app/src/main/AndroidManifest.xml
Comment (TomerPacific) [reply app/src/main/AndroidManifest.xml:133]: Sure, will do.

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (TomerPacific) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java:450]: @TheLastProject - Just so we don't run into another confusion, you would prefer the intent filters to trigger the same logic as the **Import From Filesystem button**? Meaning, the dialog opens up and let's the user select which file he/she is going to import.

A small heads up as there will be a weird flow there.
When the  Import From Filesystem button is clicked inside the application, after a user selects the type of file to import, he/she is then redirected to the file manager to choose the file.
When a file is opened from outside the application, you do not want the same user flow to happen since the user has just selected the file.

I hope that makes sense.

Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java:450]: I think what we want for the "starting outside" flow is:

1. User opens File Manager
2. User clicks file there
3. Catima opens and asks what type the import is (Catima/Stocard/Fidme/etc.)
4. User selects option and import starts with file selected in step 2

---

_Sidenote:_
I know that is slightly different from what Catima's in-app flow currently is:
1. User opens Import/Export screen
2. User clicks Import From Filesystem
3. User is asked what type the import is
4. User selects file
5. User selects option and import starts

At some point I want to change the in-app flow to:
1. User opens Import/Export screen
2. User clicks Import From Filesystem
3. User selects file
4. User is asked what type the import is
5. User selects option and import starts

But that feels out of scope for your change (although if you want to implement it you're free to, can even be a second PR)

Comment (TomerPacific) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java:450]: @TheLastProject - thanks for the feedback. Let me finish working on the changes here and I think the upgraded flow can be part of a different PR.

File: app/src/main/AndroidManifest.xml
Comment (TomerPacific) [reply app/src/main/AndroidManifest.xml]: Done.

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (TomerPacific) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java:77]: @TheLastProject,
`    private static final int PERMISSIONS_EXTERNAL_STORAGE = 1;` is just the requestCode.
It is being used here:

```
@Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == PERMISSIONS_EXTERNAL_STORAGE) {
```

Would you want me to just rename it?

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (TomerPacific) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java:456]: @TheLastProject, here is a gif recording of the "starting outside" flow:
qemu-system-x86_64_61xmebUzo3

You will notice at the end that when a user presses the back button, the application closes since it was opened by the intent filter.

Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java:456]: Yes, I noticed that's the flow and that's what I worry about. The import starts with no user interaction to make sure the user meant to import the file (maybe they just accidentally clicked the wrong app).

That's why I'm asking to add step 3, to ask the user the import type so that it will work even if the file gets renamed and to add a "confirmation step":
- Catima opens and asks what type the import is (Catima/Stocard/Fidme/etc.)

> You will notice at the end that when a user presses the back button, the application closes since it was opened by the intent filter.

That seems reasonable to me. Perhaps slightly odd, but users keeping their file manager open feels slightly odd as well, so whatever, no opinion :)

> I think the upgraded flow can be part of a different PR

Rewriting the flow that existed before this PR definitely makes sense in another PR yeah (maybe easiest to have that done first). But I'd rather not merge this PR in its current state due to the risk of users accidentally overwriting some of their data and me getting lots of angry emails about it x.x

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java:83]: My concern is about how the import seems to not care about if the storage permission is granted at all and you get this really weird situation where the import succeeds while there is a dialog asking for storage permission.

On Android 12, this manifest as follows:
https://user-images.githubusercontent.com/1885159/196546608-19def328-1210-4706-968f-735912bcb47b.mp4

On Android 6, the import doesn't seem to do anything at all (the ImportExportActivity starts, but no import is started), which might or might not be related to this. Needs more testing.

I guess this is out of scope, will block a merge but not tagging `hacktoberfest-accepted`. I need to find some time soon to test #979 in every single Android version and figure out exactly which versions need which storage permission when x.x

**Comment by TomerPacific** [comment]
@TheLastProject  - Seeing as this feature is having it's ups and downs, I think it would be preferable to define the scope of this PR and whether or not there is a need to perform any prerequisite tasks.

If we can define that, it would be much clearer for me (and for you as well).

**Comment by TheLastProject** [comment]
I see the scope as follows:

- [x] When clicking a file with the extension of a possible importable export, offer Catima as option
- [ ] Catima then asks what the type of export is -> while we can get the autodetection right in most cases, this helps prevent accidental importing if the user meant to open the file with another application but accidentally chose Catima
- [x] The export is imported
- [ ] This has to work and be tested on all supported Android versions -> this is why the "storage permission" popup worries me, as it indicates this feature starts the import before being sure it can access the files. However, see `*`

For prerequisites, I don't think any are directly necessary, but I see some related issues which will make this easier:
- [ ] Change the order of regular export importing (so: from the Catima app itself) to be first selecting the file and then selecting the type -> this way this feature can hook more easily into the existing flow
- [ ] `*` Check if the storage permission is necessary at all on all Android versions (#979) -> if the storage request was incorrectly added in the past and simply isn't necessary, the permission request can be removed from the `ImportExportActivity` and there will be no reason to worry at all about the import possibly crashing due to permission issues on some Android versions

Does this help make things more clear?

**Comment by TomerPacific** [comment]
@TheLastProject, yes this is provides a better picture, thanks.
So I will be focusing on:

- [ ] Catima then asks what the type of export is -> while we can get the autodetection right in most cases, this helps prevent accidental importing if the user meant to open the file with another application but accidentally chose Catima

I read your comments on https://github.com/CatimaLoyalty/Android/issues/979, but I am wondering if you want to go ahead and remove the storage permission logic?

**Comment by TheLastProject** [comment]
> So I will be focusing on:
> - [ ] Catima then asks what the type of export is -> while we can get the autodetection right in most cases, this helps prevent accidental importing if the user meant to open the file with another application but accidentally chose Catima

Great!

> I read your comments on https://github.com/CatimaLoyalty/Android/issues/979, but I am wondering if you want to go ahead and remove the storage permission logic?

I'm not completely sure yet as I have to test a few more Android versions, but it is looking like the storage permission logic can most likely be removed. So it seems pretty safe to ignore any issues related to it in this feature for now.

**Comment by TomerPacific** [comment]
@TheLastProject, added this dialog to the  "starting outside" flow:

qemu-system-x86_64_UGh22zSiJr

Also, why did you remove the hacktoberfest-accepted label?

**Comment by TheLastProject** [comment]
> @TheLastProject, added this dialog to the "starting outside" flow

The dialog looks fine (well, it's the same dialog that already exists), but looking at the code it seems you pretty much just copy-pasted the old dialog code so it now exists in the codebase twice? Duplicated code is really hard to maintain and makes it much easier to introduce accidental bugs so that's not a good way to go about it.

> Also, why did you remove the hacktoberfest-accepted label?

I'm not sure what you mean, this PR never had the `hacktoberfest-accepted` label. #1083 is the only one I gave that label because it seems pretty much ready to merge but because it touches the database I want to be extra careful before merging it (and deal with another unrelated-ish issue first), but that is an exception to the rule (most of the time I just merge things when they're ready).

**Comment by TheLastProject** [comment]
> Since the logic and the dialog for the in-app import flow is pretty specific, I did not know if you would want me to refactor that method entirely.

I don't understand your statement. As I said before, the intent is for the dialog and flow to be exactly the same in both situations. In both cases a file is chosen that is then imported. Catima's codebase shouldn't care if the file is selected in-app or out-of-app, it should just ask the user what type it is and import it.

I know the in-app order is currently the wrong way around (ask type first, then select file instead of first select file, then ask type) but that's why I've been suggesting possibly first fixing the in-app flow order (in a different PR) to make the out-of-app flow easier to add cleanly.

> How would you prefer to generate the dialog for the out-of-app flow?

It should use the exact same dialog as the in-app flow because both flows do the exact same thing: import a file of a supported type. The file is the same and the supported importers are the same, they should share the same flow.

**Comment by TomerPacific** [comment]
@TheLastProject  - So you are OK with the fact that when both flows share the same logic, a file imported from outside the application will open the choose file type dialog and then lead the user back to choosing a file? That is what I was trying to correct here. I understand that the current in-app flow is not in order, but I did not want to make a wrong assumption here.

**Comment by TheLastProject** [comment]
> So you are OK with the fact that when both flows share the same logic a file imported from outside the application

> will open the choose file type dialog

Yes, that's perfect.

>and then lead the user back to choosing a file?

No, it should start importing after you choose the option in the dialog. It should **not** ask to select a file again.

---

Again:
- The in-app flow is currently in the wrong order
- The desired flow is:
  1. Select file (in-app, skipped when out-of-app because the file is sent to the app through Android's "open file" system)
  2. Select type of import
  3. Import starts

Given this somehow keeps causing confusion I guess this is just really hard to do currently and we should maybe stop with this PR and first fix the order of the in-app selector before proceeding here?

**Comment by TomerPacific** [comment]
@TheLastProject , I apologize for the confusion on my end. I hope it didn't cause too much trouble or headache.

I have performed the following changes in the code base to fit your required demands:

- There is no code duplication
- The chooseImportType method now accepts a nullable Uri
- This is to provide an option in the outside of the app flow to not let the user choose a file a second time
- The regular importing from inside Catima flow has been kept the same

Let me know if this is more what you were looking for.

**Comment by TheLastProject** [comment]
That does feel nice and clean. I'll do some extra testing tomorrow (today has been quite exhausting with work and all and I'm honestly too tired to properly review and check code) but I'm quite hopeful about this iteration really hitting the nail on the head :)

**Comment by TheLastProject** [comment]
Yup, seems to work great. Thank you :)

**Comment by TomerPacific** [comment]
Thank you @TheLastProject for your patience and guidance. 
Would you like me to open an issue for fixing the importing flow like we discussed?

**Comment by TheLastProject** [comment]
Sure, that sounds good :)

**Comment by TomerPacific** [comment]
@TheLastProject  - Opened an issue #1102

**Pull request merged**
